<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kathleen Zeglinski" />

<meta name="date" content="2023-05-15" />

<title>NAb-seq bioinformatics tutorial (updated)</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="bullframe-classless.min.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">NAb-seq: Nanopore antibody sequencing</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="new_tutorial.html">Tutorial</a>
</li>
<li>
  <a href="NAb_seq_vignette.html">Old (v0.1) tutorial</a>
</li>
<li>
  <a href="Reference_guide.html">Reference sequences/databases</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">NAb-seq bioinformatics tutorial
(updated)</h1>
<h4 class="author">Kathleen Zeglinski</h4>
<h4 class="date">2023-05-15</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Welcome to the updated (v0.2) NAb-seq tutorial! This new version is
written in <a href="https://www.nextflow.io/">nextflow</a>, a workflow
language that enables portable and scalable computing. This means that
it should be really easy to use! You can find the github page <a
href="https://github.com/kzeglinski/nabseq_nf">here</a></p>
<p>A number of changes have been made since the original version. It is
recommended that you check out the <a
href="https://github.com/kzeglinski/nabseq_nf/blob/main/CHANGELOG.md">changelog</a>
for detailed information.</p>
</div>
<div id="running-the-pipeline" class="section level2">
<h2>Running the pipeline</h2>
<ol style="list-style-type: decimal">
<li>Clone the repo
(<code>git clone https://github.com/kzeglinski/nabseq_nf.git</code>)</li>
<li>Edit the <a href="#config">nextflow config file</a> to suit your
environment.</li>
<li>Make sure you have nextflow installed on your computer</li>
<li>Run the pipeline using
<code>nextflow run nabseq_nf/main.nf</code></li>
</ol>
<p>There is a nice guide on running nextflow pipelines for beginners <a
href="https://bioinformaticsworkbook.org/dataAnalysis/nextflow/01_introductionToNextFlow.html#gsc.tab=0">here</a></p>
<div id="params" class="section level3">
<h3>Parameters</h3>
<p>You can see all of the parameters using
<code>nextflow run nabseq_nf/main.nf --help</code>:</p>
<pre><code>Usage: nextflow run ./nabseq_nf/main.nf --fastq_dir [input path] --organism [organism name] --sample_sheet [sample sheet]
--help                : prints this help message

Required arguments:
--out_dir             : where the output files will be written to (default: &quot;$projectDir/results)
--fastq_dir           : where the input fastq files are located
--sample_sheet        : location of the .csv sample sheet (format: barcode01,sample_x,rat,1)

Optional (only needed for advanced users)
--barcode_dirs        : whether the input fastq files are located within folders named barcode01 etc (default: false)
--num_consensus       : maximum number of consensus sequences to generate for each sample (default: 999)
--igblast_databases   : location of the igblast databases (default: &quot;$projectDir/references/igblast/&quot;)
--reference_sequences : location of the reference sequences (default: &quot;$projectDir/references/reference_sequences/&quot;)
--trim_3p             : pattern for cutadapt to trim off the 3&#39; end (default: &quot;A{20}N{90}&quot;)
--trim_5p             : pattern for cutadapt to trim off the 5&#39; end (default: &quot;N{90}T{20}&quot;)
--medaka_model        : model to use for medaka (depends on your basecalling model, default:&quot;r941_min_sup_g507&quot;)
--report_title        : title to use for naming the report (default: &quot;NAb-seq report&quot;)</code></pre>
<p>Most of the above are self-explanatory but:</p>
<ul>
<li><p>Previously, the species/organism was specified as a command-line
argument, but this has moved to the sample sheet (so you can analyse
data from multiple different species at once). If you want to use the
built-in <code>mouse</code> or <code>rat</code> references, then you
don’t need to edit <code>--igblast_databases</code> or
<code>--reference_sequences</code>.</p></li>
<li><p><code>--barcode_dirs</code> can be used when your data is
structured like <code>reads/barcode01/{blah}.fastq</code>, as opposed to
the default NAb-seq expects, which should be like
<code>reads/barcode01{any_random_text}.fastq</code>. See the <a
href="#sample_sheet">sample sheet/file structure</a> section for more
information.</p></li>
<li><p>For the number of consensus sequences, the default is 999, which
means that NAb-seq will generate up to 999 consensus sequences for the H
chain and up to 999 consensus sequences for the light chain in a single
cell. In practice, there will never be this many and so it basically
just means <strong>generate all possible consensus sequences for each
chain in each cell</strong> (note that it is only possible to generate
consensus sequences for groups that have a count of at least 3). You can
set this number lower to save time (since the consensus calling is the
slowest part), but for hybridomas we recommend that you generate at
least the top 20 or 30 consensus sequences, given their propensity for
expressing multiple heavy and/or light chains.</p></li>
<li><p>For the adapter trimming <code>trim_3p</code> and
<code>trim_5p</code> should be <a
href="https://cutadapt.readthedocs.io/en/stable/guide.html#multiple-adapters">cutadapt
patterns</a>. You only need to change the default values if you would
like to perform trimming of custom adapter sequences. NAb-seq by default
will trim off the polyA tail.</p></li>
<li><p>For the medaka model, you can see a list of all models <a
href="https://github.com/nanoporetech/medaka/tree/master/medaka/data">here</a>.
Choose one that matches your pore number and basecalling model/guppy
version. <strong>Note: Don’t choose a model with ‘variant’ in the
name</strong></p></li>
</ul>
</div>
<div id="config" class="section level3">
<h3>Nextflow config files</h3>
<p>The nextflow config file allows you to set NAb-seq’s parameters, as
well as control other aspects of how nextflow runs this pipeline (for
example, using SLURM or AWS).</p>
<p>The config file supplied with NAb-seq has been written to run on
WEHI’s Milton HPC, and may not work on your system. You might want to
try looking at the <a href="https://github.com/nf-core/configs">nf-core
collection of institutional config files</a> to see if they have one for
your institute. Alternatively, you could try asking a nextflow expert at
your workplace!</p>
<p>For more information about writing/editing nextflow config files,
take a look at the <a
href="https://www.nextflow.io/docs/latest/config.html">official nextflow
documentation</a>.</p>
</div>
<div id="sample_sheet" class="section level3">
<h3>Sample sheet &amp; file structure</h3>
<p><a
href="https://github.com/kzeglinski/nabseq_nf/blob/main/CHANGELOG.md">As
of v0.2.2, the format of NAb-seq’s sample sheet has changed.</a> Sample
sheets should now have the structure
<code>barcode,sample_name,species,report_group</code>. For example:</p>
<pre><code>barcode,sample_name,species,report_group
barcode16,28-21-4E10-2-1,rat,1
barcode17,28-21-5H11-5-1,rat,1
barcode18,15-14-7H9-19-1,mouse,2
barcode19,42-11-A4-D1-2-1,mouse,2
barcode20,9-13-IB4-1-1,rat,3
barcode21,25-19-5A6-1-1,rat,4</code></pre>
<p>Where:</p>
<ul>
<li><p><code>barcode</code> is the nanopore barcode used.</p></li>
<li><p><code>sample_name</code> is the name of the sample. I have tested
that dashes and underscores work ok, other special characters may cause
issues so use at your own risk</p></li>
<li><p><code>species</code> is the name of the species for that sample,
either <code>rat</code> or <code>mouse</code> (<strong>case
sensitive</strong>) if you want to use NAb-seq’s built-in
references.</p></li>
<li><p><code>report_group</code> is used to create separate reports for
different groups of samples. In the example above, there would be four
reports created, one with the first two samples, one with the third and
fourth samples and then the last two samples each in their own report.
If you want all samples in the same report, just set all to 1.</p></li>
</ul>
<p>When looking for input, NAb-seq uses the sample sheet, the path
provided to <code>--fastq_dir</code> and the switch
<code>--barcode_dirs</code> to read in data from the following file
structures:</p>
<ol style="list-style-type: decimal">
<li><strong>Default behaviour:</strong> NAb-seq will look in the
<code>--fastq_dir</code> path and identify FASTQ files (ending in
.fastq, .fq, .fq.gz or .fastq.gz) that start with the barcodes in the
sample sheet (e.g. for the above example barcode16.fastq or
barcode20_pass.fq.gz or barcode18_fail.fastq.gz). If there are multiple
files for the same barcode, they will be concatenated, so <strong>it’s
not recommended to leave fail reads in the directory or they will be
used in the analysis</strong>. However, any FASTQ that don’t start with
a barcode in the sample sheet (e.g. unclassified_pass.fastq) won’t be
counted, nor will any files that don’t have a fastq file extension.</li>
<li><strong>Using the <code>--barcode_dirs</code> parameter:</strong>
NAb-seq will look in the <code>--fastq_dir</code> path for
<strong>folders</strong> that have the names of the barcodes in the
sample sheet (e.g. barcode16/ or barcode21/). Then, all FASTQ files in
each of these folders (ending in .fastq, .fq, .fq.gz or .fastq.gz) will
be concatenated. <strong>Note that file extensions can’t be mixed, so
make sure all the FASTQs are gzipped or are all unzipped otherwise the
concatenation won’t work</strong></li>
</ol>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
