<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kathleen Zeglinski" />

<meta name="date" content="2022-11-15" />

<title>Reference sequences/databases guide</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="bullframe-classless.min.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">NAb-seq: Nanopore antibody sequencing</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="new_tutorial.html">Tutorial</a>
</li>
<li>
  <a href="NAb_seq_vignette.html">Old (v0.1) tutorial</a>
</li>
<li>
  <a href="Reference_guide.html">Reference sequences/databases</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Reference sequences/databases guide</h1>
<h4 class="author">Kathleen Zeglinski</h4>
<h4 class="date">2022-11-15</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This guide will explain how to prepare reference sequences and
databases for NAb-seq (v0.2). If you are using <strong>mouse</strong> or
<strong>rat</strong> then there is no need to follow this guide, as the
references are already included within NAb-seq.</p>
</div>
<div id="all-abs" class="section level2">
<h2>Preparing the all-antibody reference file</h2>
<ol style="list-style-type: decimal">
<li><p>Visit the <a href="https://www.imgt.org/genedb/">IMGT/GENE-DB
website</a></p></li>
<li><p>Select your species, and select the ‘IG’ option under ‘Molecular
component’. Then, click submit<img src="imgt_gene_db.png" /></p>
<p>You should see a page that looks like this:</p>
<p><img src="search_results.png" /></p></li>
<li><p>To get the sequences in fasta format, scroll down to the bottom
of the page and click on the ‘select all genes’ tickbox. Then, choose
‘F+ORF+all P nucleotide sequences’ under the ‘IMGT/GENE-DB allele
reference sequences in FASTA format’ section. Finally, click submit.<img
src="getting_fasta.png" /></p>
<p>You should see a page that looks like this: <img
src="fasta_page.png" /></p></li>
<li><p>Copy the fasta format sequences (everything below the ‘Number of
results’ line, as indicated by the red box above. Paste these sequences
into a text editor (e.g. TextEdit on Mac or Notepad on Windows). Save
this file under <code>nabseq_nf/references/reference_sequences/</code>
with a .fasta extension <strong>in the format of
ORGANISM_whatever.fasta</strong> (e.g. rat_all_imgt_refs.fasta)</p></li>
<li><p>Done!</p></li>
</ol>
</div>
<div id="igblast" class="section level2">
<h2>Building IgBLAST databases</h2>
<ol style="list-style-type: decimal">
<li><p>Download reference files Visit the <a
href="https://www.imgt.org/vquest/refseqh.html#VQUEST">IMGT/V-QUEST
website</a> and scroll down until you see <strong>IMGT/GENE-DB reference
directory sets</strong>. Here, you will need to download the V, (D), J
and C genes for each of IGH, IGL and IGK. <img
src="igblast_get_ref_seqs.png" /></p></li>
<li><p>Combine reference sequences You need to combine all of the V,
(D), J and C genes into 4 files (so you should have a V genes file
containing IGHV, IGKV, IGLV and a D genes file containing IGHD, and so
on). At this stage, you may also like to concatenate the constant
sequences that come from the same gene (IMGT reports the CH1, CH2, CH3
etc of each constant region gene separately) although this is not
essential (just makes the files a bit cleaner). <strong>you should name
these files like: imgt_ORGANISM_V_unclean and imgt_ORGANISM_D_unclean
etc</strong> This is important because NAb-seq will use this naming
convention to look for your databases when running IgBLAST.</p></li>
<li><p>Clean up the reference sequences Before building the database,
you will need to <a
href="https://ftp.ncbi.nih.gov/blast/executables/igblast/release/LATEST">download
IgBLAST</a> and run <code>edit_imgt_file.pl</code> (located in the bin
directory) on each of the above files like so:
<code>ncbi-igblast-1.19.0/bin/edit_imgt_file.pl imgt_ORGANISM_V_unclean &gt; imgt_ORGANISM_V</code></p></li>
<li><p>Create the databases Finally, you need to run
<code>makeblastdb</code> like so:
<code>ncbi-igblast-1.19.0/bin/makeblastdb -parse_seqids -dbtype nucl -in imgt_ORGANISM_V</code>
If the databases don’t work, you can try checking them using the
<code>blastdbcheck</code> utility from the BLAST package. One error I
came across was missing taxonomy information, which was solved by
following the instructions at the <a
href="https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/29829-deploying-a-local-version-of-blast-having-issues">bottom
of this page</a>. <strong>important</strong> make sure that these
databases are located in
<code>nabseq_nf/references/igblast/databases/</code></p></li>
</ol>
</div>
<div id="igblast-internal" class="section level2">
<h2>IgBLAST internal &amp; optional data</h2>
<p>If you are using one of the organisms supported by IgBLAST (human,
mouse, rat, rabbit and rhesus_monkey), you can get these files from
within the standalone IgBLAST program. Guide for setting up custom
organisms coming soon!</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
